src/qiboml/models/abstract.py

```diff
@@ -0,0 +1,67 @@
+"""Defines the general structure of a qiboml module"""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+from qibo import Circuit
+from qibo.config import raise_error
+from qibo.gates import abstract
+
+from qiboml.backends import TensorflowBackend as JaxBackend
+
+
+@dataclass
+class QuantumCircuitLayer(ABC):
+
+    nqubits: int
+    qubits: list[int] = None
+    circuit: Circuit = None
+    initial_state: "ndarray" = None
+    backend: "qibo.backends.Backend" = JaxBackend()
+
+    def __post_init__(self) -> None:
+        if self.qubits is None:
+            self.qubits = range(self.nqubits)
+        self.circuit = Circuit(self.nqubits)
+
+    @abstractmethod
+    def forward(self, x):
+        pass
+
+    @abstractmethod
+    def backward(self, input_grad: "ndarray"):
+        pass
```

Comment:
Why `backward` should differ from `forward`? I can see that in simulation you could have both, but that does not extend to hardware. Is it truly worth to keep both?

Reply:
The forward and backward pass in ML are two distinct steps. In the forward you feed the data and process them layer by layer until you compute the output of the network, this would be the execution of the circuit in our case. Whereas, the backward pass involves the gradient calculation and aggregation through the chain rule, starting from the output layer and going backward till the input layer, backpropagation indeed. I would rather keep them separated, for the hardware execution case the backward will simply involve several calls of the forward with different parameters.

Reply:
This is not always the case, as backward gradient accumulation is only one possible way of computing the gradient, but forward mode is also possible, and convenient according to the problem.
https://en.wikipedia.org/wiki/Automatic_differentiation#Forward_and_reverse_accumulation
(the trade-off is between the number of inputs and outputs, i.e. the Jacobian shape)

If you want to keep function evaluation separate from the gradient one is perfectly fine, but I'd choose at least different names from *forward* and *backward*, and what I was saying is that most likely you will need a single name for the gradient (i.e. we don't want to implement evaluation + backward-gradient + forward-gradient, but just evaluation + gradient).